lib/pharos/ssh/client.rb

```diff
@@ -134,17 +146,34 @@ def interactive_session
       end
 
       def connected?
-        synchronize { @session && !@session.closed? }
+        synchronize { !session.nil? && !session.closed? }
+      end
+
+      def gateway_shutdown
+        synchronize do
+          return unless @gateway
+
+          @gateway.shutdown!
+          sleep 0.1 until !@gateway.active?
+          @gateway = nil
+        end
       end
 
       def disconnect
-        synchronize { @session.close if @session && !@session.closed? }
+        synchronize do
+          session.close if session && !session.closed?
+          bastion.host.ssh.gateway_shutdown if bastion
+          gateway_shutdown
+          @session = nil
+        end
       end
 
       private
 
       def require_session!
-        raise Error, "Connection not established" if @session.nil? || @session.closed?
+        connect(timeout: 3) unless connected?
```

Comment:
Changed the "SSH::Client.exec` etc that use `require_session!` attempt to make a connection instead of failing directly.